net/HttpRequest.hpp

```diff
@@ -1140,12 +1149,18 @@ class Session final : public ProtocolHandlerInterface
         {
             // Remove consumed data.
             data.erase(data.begin(), data.begin() + read);
+            close = !isConnected();
```

Comment:
Why do we introduce this extra method / channel here to show that the socket is closed ? what is wrong with returning -1 from readData?
Why do we need to handle the disconnect this time around ? we can always spin the loop again if we read a chunk of data, and ? - I'm really missing a rationale here. I guess it is at least not some std::pair<> we're returning ;->

Reply:
Still - it's all isolated to the http code at least ;-)

Reply:
Returning -1 will imply an error (we won't remove the read data and will simply close the connection). That's not what we want here. We have consumed the full data and there is nothing more to do. We just need to close the connect and release the socket, but *only* if the other side does not want to reuse the socket. So this is a condition that we learn about from the header, which is something that we get after parsing it.

Sorry if the commit log is didn't expound on all this. Perhaps it's not obvious that we should maintain the connection in non-error cases by default (it's a requirement of the RFC for HTTP/1.1).

Reply:
Ah! - ok, makes perfect sense =) Of course - we can store up the completion information and just return that state on the next readData - surely (?) there's no need for another flag. Just to check - we are coping with the case where the poll returns not only a chunk of data, but also a connection-closed flag - so we process all the data before closing (?) =)

Reply:
@mmeeks Yes, that is what we do: we tell poll to close the connection (and remove our socket etc.). Unfortunately, it seems our poll is prone to invoking handleIncomingMessage even after we have set the disposition to Closed. That was the issue fixed here. The `if (!isConnected())` at the top of the function is to protect against rentrancy after closing and releasing the socket member object (which was segfaulting).

Also worth mentioning that there is *no* new state: the `_connected` member had been there because it's used to reconnect upon re-use (or indeed to connect the very first time an http::Session is used).